KAFKA-18269: Remove deprecated protocol APIs support (KIP-896, KIP-724)

ijuma opened the pull request:
Included in this change:
1. Remove deprecated protocol api versions from json files.
3. Remove fields that are no longer used from json files (affects ListOffsets, OffsetCommit, DescribeConfigs).
4. Remove record down-conversion support from KafkaApis.
5. No longer return `Errors.UNSUPPORTED_COMPRESSION_TYPE` on the fetch path[1].
6. Deprecate `TopicConfig. MESSAGE_DOWNCONVERSION_ENABLE_CONFIG` and made the relevant
configs (`message.downconversion.enable` and `log.message.downcoversion.enable`) no-ops since
down-conversion is no longer supported. It was an oversight not to deprecate this via KIP-724.
7. Fix `shouldRetainsBufferReference` to handle null request schemas for a given version.
8. Simplify producer logic since it only supports the v2 record format now.
9. Fix tests so they don't exercise protocol api versions that have been removed.
10. Add upgrade note.

Testing:
1. System tests have a lot of failures, but those tests fail for trunk too and I didn't see any issues specific to this change - it's hard to be sure given the number of failing tests, but let's not block on that given the other testing that has been done (see below).
3. Java producers and consumers with version 0.9-0.10.1 don't have api versions support and hence they fail in an ungraceful manner: the broker disconnects and the clients reconnect until the relevant timeout is triggered.
4. Same thing seems to happen for the console producer 0.10.2 although it's unclear why since api versions should be supported. I will look into this separately, it's unlikely to be related to this PR.
5. Console consumer 0.10.2 fails with the expected error and a reasonable message[2].
6. Console producer and consumer 0.11.0 works fine, newer versions should naturally also work fine.
7. kcat 1.5.0 (based on librdkafka 1.1.0) produce and consume fail with a reasonable message[3][4].
8. kcat 1.6.0-1.7.0 (based on librdkafka 1.5.0 and 1.7.0 respectively) consume fails with a reasonable message[5].
9. kcat 1.6.0-1.7.0 produce works fine.
10. kcat 1.7.1  (based on librdkafka 1.8.2) works fine for consumer and produce.
11. confluent-go-client (librdkafka based) 1.8.2 works fine for consumer and produce.
12. I will test more clients, but I don't think we need to block the PR on that.

Note that this also completes part of KIP-724: produce v2 and lower as well as fetch v3 and lower are no longer supported.

Future PRs will remove conditional code that is no longer needed (some of that has been done in KafkaApis,
but only what was required due to the schema changes). We can probably do that in master only as it does
not change behavior.

Note that I did not touch `ignorable` fields even though some of them could have been
changed. The reasoning is that this could result in incompatible changes for clients
that use new protocol versions without setting such fields _if_ we don't manually
validate their presence. I will file a JIRA ticket to look into this carefully for each
case (i.e. if we do validate their presence for the appropriate versions, we can
set them to ignorable=false in the json file).

[1] We would return this error if a fetch < v10 was used and the compression topic config was set
to zstd, but we would not do the same for the case where zstd was compressed at the producer
level (the most common case). Since there is no efficient way to do the check for the common
case, I made it consistent for both by having no checks.
[2] ```org.apache.kafka.common.errors.UnsupportedVersionException: The broker is too new to support JOIN_GROUP version 1```
[3]```METADATA|rdkafka#producer-1| [thrd:main]: localhost:9092/bootstrap: Metadata request failed: connected: Local: Required feature not supported by broker (0ms): Permanent```
[4]```METADATA|rdkafka#consumer-1| [thrd:main]: localhost:9092/bootstrap: Metadata request failed: connected: Local: Required feature not supported by broker (0ms): Permanent```
[5] `ERROR: Topic test-topic [0] error: Failed to query logical offset END: Local: Required feature not supported by broker`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma committed:
Remove deprecated protocol api versions.

ijuma committed:
`shouldRetainsBufferReference` should handle the case where the request schema is null for a given version

ijuma committed:
Simplify producer logic since it only supports the v2 record format now

ijuma committed:
Fix tests

ijuma committed:
Fix integration tests

ijuma committed:
FindCoordinatorResponse was missed

ijuma committed:
Remove fields that no longer exist in the supported versions

ijuma committed:
Adjust code and tests to handle the removed fields and protocol versions

ijuma committed:
Fix more tests

ijuma committed:
Fix test failures

ijuma committed:
Upgrade note and deprecate "message.downconversion.enable"

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -834,81 +832,6 @@ class KafkaApis(val requestChannel: RequestChannel,
-
-      if (logConfig.exists(_.compressionType == BrokerCompressionType.ZSTD) && versionId < 10) {
-        trace(s"Fetching messages is disabled for ZStandard compressed partition $tp. Sending unsupported version response to $clientId.")
-        FetchResponse.partitionResponse(tp, Errors.UNSUPPORTED_COMPRESSION_TYPE)
```
I removed this check because it's kind of pointless (this topic config is rarely used and we never checked for the most common case of compression happening on the producer) and added unnecessarily complexity once down conversion was removed. Let me know if this is reasonable.

mumrah replied:
IIUC, we aren't doing on-the-fly down conversion anymore since we've dropped all of Fetch RPC versions that didn't support record version 2. 

And for extant data with the older record versions, they are forward compatible with any newer Fetch RPC. 

Is that right?

ijuma replied:
> IIUC, we aren't doing on-the-fly down conversion anymore since we've dropped all of Fetch RPC versions that didn't support record version 2.

Correct.

> And for extant data with the older record versions, they are forward compatible with any newer Fetch RPC.

Correct, KIP-724 explains this in more detail. Unfortunately, the consumers need to support all record versions for now. See "Future Work" for more on what we would have to do to change that.

mumrah commented on clients/src/main/resources/common/message/CreateTopicsRequest.json:
```diff
@@ -18,6 +18,8 @@
   "type": "request",
   "listeners": ["zkBroker", "broker", "controller"],
   "name": "CreateTopicsRequest",
+  // Versions 0-1 were removed in Apache Kafka 4.0, Version 2 is the new baseline.
```
In the KIP, it states

> CreateTopics: V0-V1 due to clients (Sarama) support – V0-V2 per the baseline

Seems we are removing version 0 and 1. What does "V0-V2 per the baseline" mean for this and other RPCs?

ijuma replied:
It can be ignored, that is what would have happened if we had used 2.1 as the baseline without the adjustment for third party clients (think of it as a way to describe how we got to the V0-V1 as the final answer).

mumrah commented on clients/src/main/resources/common/message/FindCoordinatorRequest.json:
```diff
@@ -18,6 +18,8 @@
   "type": "request",
   "listeners": ["zkBroker", "broker"],
   "name": "FindCoordinatorRequest",
+  // Versions 0 was removed in Apache Kafka 4.0, Version 1 is the new baseline.
+  //
```
KIP states

> FindCoordinator: none due to clients (Sarama, kafka-python) support – V0-V1 per the baseline

Does this mean we should not be removing any versions?

ijuma replied:
Yes, good catch. It looks like this is a second case where we deprecated something incorrectly. I checked the KIP and this was indeed an update:

https://cwiki.apache.org/confluence/pages/diffpagesbyversion.action?pageId=235838251&selectedPageVersions=18&selectedPageVersions=17

mumrah commented on clients/src/main/resources/common/message/DescribeConfigsResponse.json:
```diff
@@ -17,10 +17,11 @@
   // Version 1 adds ConfigSource and the synonyms.
   // Starting in version 2, on quota violation, brokers send out responses before throttling.
   // Version 4 enables flexible versions.
-  "validVersions": "0-4",
```
Should we remove the `IsDefault` field since it is version 0 only?

ijuma replied:
Good point, this hadn't been mentioned in the changelog and I missed it.

mumrah reviewed (commented):
@ijuma did a pass through the schemas. 

I had one question about the wording in the KIP. We have several RPCs that say something like:

> DescribeConfigs: V0 due to clients (librdkafka) support – V0-V1 per the baseline

What does the "V0-V1 per the baseline" refer to? Is that the supported version range of the baseline Kafka version 2.1? 

Also, now that I've started reviewing, can you try to avoid force pushes? Otherwise it will reset the "Viewed" state in my review which makes it hard to see what has been reviewed already.

ijuma commented:
> Also, now that I've started reviewing, can you try to avoid force pushes? Otherwise it will reset the "Viewed" state in my review which makes it hard to see what has been reviewed already.

Totally, I usually don't force push once someone starts reviewing. Even in this PR, I haven't actually rewritten history (I added commits) except for one commit I pushed by mistake. It is interesting that the `-f` flag is still captured even when the push doesn't require it. Will make sure I don't use the latter from now on.

ijuma committed:
Remove DescribeConfigsResponse.IsDefault and v0 specific code in constructor

ijuma committed:
Restore FindCoordinator versions

mumrah commented on clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java:
```diff
@@ -42,27 +42,16 @@
 public class ProduceRequest extends AbstractRequest {
     public static final short LAST_STABLE_VERSION_BEFORE_TRANSACTION_V2 = 11;
 
-    public static Builder forMagic(byte magic, ProduceRequestData data, boolean useTransactionV1Version) {
```
Nice to see this record version stuff going away 😄

mumrah commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -344,8 +343,7 @@ public RecordAppendResult append(String topic,
                 if (buffer == null) {
-                    byte maxUsableMagic = apiVersions.maxUsableProduceMagic();
-                    int size = Math.max(this.batchSize, AbstractRecords.estimateSizeInBytesUpperBound(maxUsableMagic, compression.type(), key, value, headers));
+                    int size = Math.max(this.batchSize, AbstractRecords.estimateSizeInBytesUpperBound(RecordBatch.CURRENT_MAGIC_VALUE, compression.type(), key, value, headers));
```
nit: can you break this long line?

mumrah commented on clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java:
```diff
@@ -225,33 +214,28 @@ public void clearPartitionRecords() {
         data = null;
     }
 
+    //FIXME Check this logic
```
Is this still valid?

ijuma replied:
This seems fine, I'll just remove the FIXME.

mumrah reviewed (commented):
Took a pass through the rest of the PR. A few comments inline. Overall, I think it looks good. I think we'll be leaning on the test suite for this change :) 

Will we be able to delete LazyDownConversionRecords and the remaining down conversion code after this PR lands?

ijuma commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -379,7 +378,7 @@ public void testSerializationSpecialCases() {
         checkOlderFetchVersions();
         // Metadata
-        checkRequest(MetadataRequest.Builder.allTopics().build((short) 2));
+        checkRequest(MetadataRequest.Builder.allTopics().build((short) 4)); //FIXME Does version 4 have the special cases?
```
I am going through the FIXMEs and will push a commit with the fixes soon.

ijuma replied:
This seems fine, version 4 only added `AllowAutoTopicCreation` - it should have whatever special case version 2 had (not ideal that we didn't document that).

ijuma commented:
> Will we be able to delete LazyDownConversionRecords and the remaining down conversion code after this PR lands?

Yes. This and a few other things are what I meant by:

> Future PRs will remove conditional code that is no longer needed (some of that has been done in KafkaApis,
but only what was required due to the schema changes). We can probably do that in master only as it does
not change behavior.

> I think we'll be leaning on the test suite for this change :)

Yeah, I think that will ensure the remaining protocol versions work correctly. For the clients that are no longer supported, I will test them manually and considering what we can do via the system tests.

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kip-896-remove-old-protocol-versions

* apache-github/trunk: (25 commits)
  KAFKA-18270: FindCoordinator v0 incorrectly tagged as deprecated (#18262)
  KAFKA-18284: Add group coordinator records for Streams rebalance protocol (#18228)
  MINOR: Fix flaky state updater test (#18253)
  MINOR: improve StreamsResetter logging (#18237)
  KAFKA-18227: Ensure v2 partitions are not added to last transaction during upgrade (#18176)
  Add IT for share consumer with duration base offet auto reset (#18251)
  KAFKA-18283: Add StreamsGroupDescribe RPC definitions (#18230)
  KAFKA-18241: add docs check to CI (#18183)
  KAFKA-18223 Improve flaky test report (#18212)
  MINOR Remove triage label in nightly job (#18147)
  KAFKA-18294 Remove deprecated SourceTask#commitRecord (#18260)
  KAFKA-18264 Remove NotLeaderForPartitionException (#18211)
  KAFKA-13722: Refactor SerdeGetter (#18242)
  KAFKA-18094 Remove deprecated TopicListing(String, Boolean) (#18248)
  KAFKA-18282: Add StreamsGroupHeartbeat RPC definitions (#18227)
  KAFKA-18026: KIP-1112 migrate KTableSuppressProcessorSupplier (#18150)
  KAFKA-18026: transition KTable#filter impl to use processor wrapper (#18205)
  KAFKA-18293 Remove `org.apache.kafka.common.security.oauthbearer.secured.OAuthBearerLoginCallbackHandler` and `org.apache.kafka.common.security.oauthbearer.secured.OAuthBearerValidatorCallbackHandler` (#18244)
  MINOR: add assertion about groupEpoch and targetAssignmentEpoch to testConsumerGroups (#18203)
  KAFKA-17960; PlaintextAdminIntegrationTest.testConsumerGroups fails with CONSUMER group protocol (#18234)
  ...

ijuma committed:
Remove FIXMEs

ijuma commented:
@mumrah I merged master to the branch, fixed the FIXMEs (simply removed them) and updated the PR description with the testing I have done so far and what's in progress. I expect the PR is mostly ready, the only thing I intend to do before merging is fixing any system test failures due to this PR. Let me know if you think there's anything else that needs to be done here.

ijuma committed:
Restore produce v4-v6 (inclusive) to workaround librdkafka bug

ijuma committed:
Fix long line

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kip-896-remove-old-protocol-versions

* apache-github/trunk:
  KAFKA-18312: Added entityType: topicName to SubscribedTopicNames in ShareGroupHeartbeatRequest.json (#18285)
  HOTFIX: fix incompatible types: Optional<TimestampAndOffset> cannot be converted to Option<TimestampAndOffset> (#18284)
  MINOR Fix some test-catalog issues (#18272)
  KAFKA-18180: Move OffsetResultHolder to storage module (#18100)
  KAFKA-18301; Make coordinator records first class citizen (#18261)
  KAFKA-18262 Remove DefaultPartitioner and UniformStickyPartitioner (#18204)
  KAFKA-18296 Remove deprecated KafkaBasedLog constructor (#18257)
  KAFKA-12829: Remove old Processor and ProcessorSupplier interfaces (#18238)
  KAFKA-18292 Remove deprecated methods of UpdateFeaturesOptions (#18245)
  KAFKA-12829: Remove deprecated Topology#addProcessor of old Processor API (#18154)
  KAFKA-18035, KAFKA-18306, KAFKA-18092: Address TransactionsTest flaky tests (#18264)
  MINOR: change the default linger time in the new coordinator (#18274)
  KAFKA-18305: validate controller.listener.names is not in inter.broker.listener.name for kcontrollers (#18222)
  KAFKA-18207: Serde for handling transaction records (#18136)
  KAFKA-13722: Refactor Kafka Streams store interfaces (#18243)
  KAFKA-17131: Refactor TimeDefinitions (#18241)
  MINOR: Fix MessageFormatters (#18266)
  Mark flaky tests for Dec 18, 2024 (#18263)

ijuma commented:
@mumrah I think this is ready to go. I updated the PR description with the client versions I tested. One of the consequences of the testing was KAFKA-18334. I adjusted this PR to take it into account and submitted #18288 as well.

mumrah reviewed (approved):
Thanks for the work on this @ijuma, LGTM.

Will you merge #18288 in first?

ijuma commented:
Yes @mumrah, that's the plan. And then merge trunk to this PR and then merge this PR once green.

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into kip-896-remove-old-protocol-versions

* apache-github/trunk:
  KAFKA-18334: Produce v4-v6 should be undeprecated (#18288)
  KAFKA-13722: code cleanup after deprecated StateStore.init() was removed (#18249)
  KAFKA-15370: Support Participation in 2PC (KIP-939) (1/N) (#17687)

ijuma committed:
Adjust test now that the min produce request supported is v3

ijuma committed:
Adjust test given type change due to merge from master

frankvicky commented:
Hi @ijuma,

I noticed a discrepancy between this PR and [KIP-896](https://cwiki.apache.org/confluence/x/K5sODg) regarding `DeleteGroups` support:
```
DeleteGroups: none due to clients (Sarama) support - V0 per the baseline
* librdkafka: [V1](https://github.com/confluentinc/librdkafka/blob/v1.8.2/src/rdkafka_request.c#L4147)
* KafkaJS: [V1](https://github.com/tulios/kafkajs/blob/v1.15.0/src/protocol/requests/deleteGroups/v1/request.js#L7)
* Sarama: [V0](https://github.com/Shopify/sarama/blob/v1.29.1/describe_groups_request.go#L21)
* kafka-python: [V1](https://github.com/dpkp/kafka-python/blob/2.0.2/kafka/admin/client.py#L1328)
```

According to the KIP, Sarama only supports V0 of the `DeleteGroups RPC`. However, I found issue (https://github.com/IBM/sarama/issues/2408) which suggests Sarama will upgrade its RPC support. (And it has support v1 after https://github.com/IBM/sarama/pull/2551) If this is the case, should we update the description in KIP-896 (https://cwiki.apache.org/confluence/x/K5sODg)?

WDYT?

ijuma commented:
No, that's not a discrepancy as far as I can tell. The KIP specifies a particular version of Sarama and links to the relevant code. Can you please check again with that in mind?

frankvicky commented:
Oh, I should check trunk first.
Thanks for pointing out 🙇🏼

ijuma commented:
Thanks for checking - it's helpful to make sure!
